### ISSUE TITLE
Vector::is_inline

### ISSUE THREAD

**Issue by cmyr** [bodil/im-rs#129]
I have a particular use case: I am looking at adopting `im` collection types as the blessed way of doing collections in [druid](https://github.com/xi-editor/druid). For this to work, I need to be able to implement our `Data` trait, which is a fast & loose equality check. This will generally be a matter of just calling `ptr_eq` on the collections.

This is a bit weird with `Vector`, since `ptr_eq` returns false for all inline collections. In the inline case, I would like to fall back to item-wise equality checking, but to do that I need to know if we're inline.

I could come up with something that would probably work, based on `len()` and my element size, but it would be nice if I had a robust way of determining whether a `Vector` was inline, so I could know whether or not I should fall back on `eq`.

tldr; would you accept a patch that adds a fn `is_inline(&self) -> bool` to `Vector`?

Bonus question: I haven't looked much at the other collections yet; but if there are other places I would encounter this problem, it might be good to talk about those now as well.

Docs look great, code looks great, hope we can make this work!

**Comment by bodil** [comment]
I'm not opposed to an `is_inline()` method, but are you sure you need it? I mean, if `ptr_eq` returns false you'll have to revert to `eq` to decide equality whether or not it's inline, or am I missing something?

The other datatypes don't do inlining, so this issue is unique to `Vector`.

**Comment by xStrom** [comment]
It's mostly about fresh clones and having to do a ton of work if the collection might have changed.

Druid does a lot of cloning without necessarily changing anything. It compares the clones to see if they are guaranteed to still be the same. Currently that means that small collections always act as if they've changed, which ends up in a lot of song and dance for nothing.

If we could know that it's inline (a small collection) then we could do equality checks in that case to avoid the big work. We would never do it for large collections, because fresh clones would return `true` from `ptr_eq` which is what we need.

**Comment by bodil** [comment]
Makes sense. I'll add an `is_inline()` method.

**Comment by cmyr** [comment]
@bodil thank you for being so responsive to our esoteric needs!

**Comment by bodil** [comment]
What would even be the point of a persistent data structure library if it isn't useful for storing UI state? 😊

**Comment by cmyr** [comment]
Careful what you wish for, you might get me back in a couple weeks blathering about diffs. 😬

**Comment by bodil** [comment]
Tagged a new release, fyi, I guess it was about time.

**Comment by cmyr** [comment]
thanks!
